### PR TITLE
InvalidLinkBear: Improve documentation

### PR DESCRIPTION
**Pull request by sils1297**

### CODE REVIEW

File: bears/general/InvalidLinkBear.py [hunk 1]
```diff
@@ -40,7 +40,10 @@ def find_links_in_file(file, timeout):
 
     def run(self, filename, file, timeout: int=DEFAULT_TIMEOUT):
         '''
-        Yields results for all invalid links in a file.
+        Finds links in any text file and checks if they are valid.
```
Comment (tushar-rishav) [inline bears/general/InvalidLinkBear.py]: Should rather adopt Imperatives here? and other places too @sils1297

Comment (sils1297) [inline bears/general/InvalidLinkBear.py]: ah for bears I don't know...

Comment (sils1297) [inline bears/general/InvalidLinkBear.py]: we haven't specified it yet :P

**Comment by sils1297** [comment]
@tushar-rishav updated

**Comment by Adrianzatreanu** [comment]
ack 98c3f1e

**Comment by Adrianzatreanu** [comment]
ack 6a2ac8a

File: bears/c_languages/ClangComplexityBear.py [hunk 2]
```diff
@@ -85,10 +85,20 @@ def run(self, filename, file, max_complexity: int=8):
                 affected_code = (SourceRange.from_clang_range(cursor.extent),)
                 yield Result(
                     self,
-                    "The cyclomatic complexity of function {function} is "
-                    "{complexity} which exceeded maximal recommended value "
+                    "This function {function} should be simplified. The "
```
Comment (Adrianzatreanu) [inline bears/c_languages/ClangComplexityBear.py]: i don't think this sounds good. You can't quite say: ``this function sum`` for example

File: bears/c_languages/ClangComplexityBear.py [hunk 3]
```diff
@@ -85,10 +85,20 @@ def run(self, filename, file, max_complexity: int=8):
...
                         complexity=complexity,
                         rec_value=max_complexity),
-                    affected_code=affected_code)
+                    affected_code=affected_code,
+                    additional_info=(
+                        "The cyclomatic complexity is a metric that measures "
+                        "how complicated a function is by counting branches "
+                        "and exits of each function.\n\n"
+                        "Your function seems to be complicated and should be "
+                        "refactored so it can be understood by other people "
```
Comment (Adrianzatreanu) [inline bears/c_languages/ClangComplexityBear.py]: so **that**

File: bears/general/SpaceConsistencyBear.py [hunk 4]
```diff
@@ -14,7 +14,9 @@ def run(self,
             tab_width: int=SpacingHelper.DEFAULT_TAB_WIDTH,
             enforce_newline_at_EOF: bool=True):
         '''
-        Checks the space consistency for each line.
+        Check and correct spacing for all textual data. This includes usage of
+        tabs vs. spaces, trailing whitespace and (missing) newlines before
+        the end of the file.
```
Comment (Adrianzatreanu) [inline bears/general/SpaceConsistencyBear.py]: do you think it's worth adding the link from stackoverflow regarding newlines before the EOF here?

**Comment by Adrianzatreanu** [comment]
10ed0de needs work

File: bears/c_languages/ClangComplexityBear.py (hunk 2 above)
Comment (tushar-rishav) [inline bears/c_languages/ClangComplexityBear.py]: Maybe `The function sum` sounds better? @Adrianzatreanu

Comment (Adrianzatreanu) [inline bears/c_languages/ClangComplexityBear.py]: maybe even without ``the``

Comment (Adrianzatreanu) [inline bears/c_languages/ClangComplexityBear.py]: try "Function sum should be simplified."

File: bears/general/SpaceConsistencyBear.py (hunk 4 above)
Comment (tushar-rishav) [inline bears/general/SpaceConsistencyBear.py]: IMO not. A simple google search might provide that link if a user *really* cares.

Comment (Adrianzatreanu) [inline bears/general/SpaceConsistencyBear.py]: well i agree with you, but he provided a link @ https://github.com/coala-analyzer/coala-bears/pull/333/commits/10ed0de7dfc1dcb44390696ebc4aca47423c2606 , that's why i wanted to keep it consistent. should we have link on both, or neither imo

Comment (sils1297) [inline bears/general/SpaceConsistencyBear.py]: I think the additional_info should be as extensive as possible and also provide links to additional resources

File: bears/c_languages/ClangComplexityBear.py (hunk 3 above)
Comment (sils1297) [inline bears/c_languages/ClangComplexityBear.py]: isn't both ok?

File: bears/general/SpaceConsistencyBear.py (hunk 4 above)
Comment (Adrianzatreanu) [inline bears/general/SpaceConsistencyBear.py]: so do we want a link here or not?

File: bears/c_languages/ClangComplexityBear.py (hunk 3 above)
Comment (Adrianzatreanu) [inline bears/c_languages/ClangComplexityBear.py]: well, in my mind only ``so that`` sounds good, but they might be both correct :P

File: bears/general/SpaceConsistencyBear.py (hunk 4 above)
Comment (sils1297) [inline bears/general/SpaceConsistencyBear.py]: not here but ideally in additional_info

File: bears/c_languages/ClangComplexityBear.py (hunk 2 above)
Comment (sils1297) [inline bears/c_languages/ClangComplexityBear.py]: added quotes, refusing to remove the "the" because IMO it sucks without it :P

File: bears/c_languages/ClangComplexityBear.py [hunk 5]
```diff
@@ -85,10 +85,20 @@ def run(self, filename, file, max_complexity: int=8):
...
                         rec_value=max_complexity),
-                    affected_code=affected_code)
+                    affected_code=affected_code,
+                    additional_info=(
+                        "The cyclomatic complexity is a metric that measures "
+                        "how complicated a function is by counting branches "
+                        "and exits of each function.\n\n"
+                        "Your function seems to be complicated and should be "
+                        "refactored so that it can be understood by other "
+                        "people easily.\n\nSee "
```
Comment (AbdealiJK) [inline bears/c_languages/ClangComplexityBear.py]: Out of curiosity ... does using`\n` look fine in windows ?

I remember C++ has `std::endl` because the end-line character is different in some systems.

File: bears/general/SpaceConsistencyBear.py [hunk 6]
```diff
@@ -41,11 +42,21 @@ def run(self,
...
+                        "give you more information about why this may or may "
+                        "not be important to you.")
 
             if not allow_trailing_whitespace:
                 replacement = replacement.rstrip(" \t\n") + "\n"
                 if replacement != line.rstrip("\n") + "\n":
                     result_texts.append("Trailing whitespaces.")
+                    additional_info_texts.append(
+                        "Your source code contains trailing whitespaces. Those "
+                        "usually have no meaning and thus are redundant and "
```
Comment (AbdealiJK) [inline bears/general/SpaceConsistencyBear.py]: This message seems to be repetitive because you say no meaning *and* redundant.

`Your source code contains trailing whitespaces which usually have no meaning and should be removed.` ?

Comment (AbdealiJK) [inline bears/general/SpaceConsistencyBear.py]: Also instead of "should be removed", do "you may want to consider removing them" ?

File: bears/vcs/git/GitCommitBear.py [hunk 7]
```diff
@@ -27,7 +27,7 @@ def run(self,
             shortlog_trailing_period: bool=None,
             shortlog_imperative_check: bool=True):
         """
-        Checks the current git commit message at HEAD.
+        Check the current git commit message at HEAD.
```
Comment (AbdealiJK) [inline bears/vcs/git/GitCommitBear.py]: Are we doing this everywhere ? Can we use pydocstyle then ?
Loads of places where this is not imperative

File: bears/c_languages/ClangComplexityBear.py (hunk 5 above)
Comment (sils1297) [inline bears/c_languages/ClangComplexityBear.py]: it should, pythons `print` function takes care of such differences so we don't have to

Comment (sils1297) [inline bears/c_languages/ClangComplexityBear.py]: it should, pythons `print` function takes care of such differences so we don't have to

File: bears/vcs/git/GitCommitBear.py (hunk 7 above)
Comment (sils1297) [inline bears/vcs/git/GitCommitBear.py]: /me feels fresh air surrounding his head.

It's so great to have you back. :)

File: bears/c_languages/ClangComplexityBear.py (hunk 5 above)
Comment (AbdealiJK) [inline bears/c_languages/ClangComplexityBear.py]: awesome

File: bears/general/InvalidLinkBear.py [hunk 8]
```diff
@@ -40,7 +40,10 @@ def find_links_in_file(file, timeout):
 
     def run(self, filename, file, timeout: int=DEFAULT_TIMEOUT):
         '''
-        Yields results for all invalid links in a file.
+        Find links in any text file and check if they are valid.
+
+        The bear will check if the page exists and propose an automatic fix if
+        it is a redirect.
```
Comment (SanketDG) [inline bears/general/InvalidLinkBear.py]: Add "If the page is not found or its a server side error, then those links will be displayed too"

File: bears/general/SpaceConsistencyBear.py (hunk 6 above)
Comment (Adrianzatreanu) [inline bears/general/SpaceConsistencyBear.py]: yeah, i totally agree with @AbdealiJK on this one: a formulation i found:
``Your source code contains trailing whitespaces which may have no meaning and you should consider removing them.``
damn, when a program is giving you advice.

**Comment by Adrianzatreanu** [comment]
18ec9ac needs work imo

File: bears/general/SpaceConsistencyBear.py [hunk 9]
```diff
@@ -41,11 +42,21 @@ def run(self,
                 if replacement[-1] != "\n":
                     replacement += "\n"
                     result_texts.append("No newline at EOF.")
+                    additional_info_texts.append(
+                        "A trailing newline character ('\\n') is missing from "
+                        "your file. "
+                        "<http://stackoverflow.com/a/5813359/3212182> will "
+                        "give you more information about why this may or may "
+                        "not be important to you.")
```
Comment (Adrianzatreanu) [inline bears/general/SpaceConsistencyBear.py]: too long here i think. what about ``gives you more information on why you would need one.``

- change to present tense
- it's a little shorter

File: tests/c_languages/ClangComplexityBearTest.py [hunk 10]
```diff
@@ -59,11 +59,14 @@ def test_output(self):
             end_column=2),)
         expected_result = Result(
             self.bear,
-            "The cyclomatic complexity of function levels(int, int, int) is "
-            "10 which exceeded maximal recommended value of 8.",
+            "This function '{function}' should be simplified. The cyclomatic "
```
Comment (Adrianzatreanu) [inline tests/c_languages/ClangComplexityBearTest.py]: im sorry, i still disagree on using this. ``This function sum`` sounds bad. Leaving ``Function {function}`` would be waaay better imo

File: bears/vcs/git/GitCommitBear.py (hunk 7 above)
Comment (sils1297) [inline bears/vcs/git/GitCommitBear.py]: do we have a bear/issue for this? This is out of scope for this PR of course.

Comment (AbdealiJK) [inline bears/vcs/git/GitCommitBear.py]: yep, out of scope for this PR

If this is a new codestyle we're introducing then we need to make an issue to do this and add documentation appropriately in codestyle docs.

Not sure if there was some discussion about adding this to our codestyle (Didn't find any issue)

File: tests/c_languages/ClangComplexityBearTest.py (hunk 10 above)
Comment (sils1297) [inline tests/c_languages/ClangComplexityBearTest.py]: changed to:

> The function 'levels(int, int, int)' should be simplified. Its cyclomatic complexity is 10 which exceeds maximal recommended value of 8.

**Comment by sils1297** [comment]
reack 13e433487b5893f9f0a38536b34044b0b01f4dfe

**Comment by sils1297** [comment]
rebased this on #343 it'll be rebased again whan #343 is completed, currently blocked through the codecov deployment. Do not review the WIP commit on top but it makes totally sense to review the rest!

File: bears/general/InvalidLinkBear.py [hunk 11]
```diff
@@ -39,11 +39,15 @@ def find_links_in_file(file, timeout):
                 yield line_number + 1, link, code
 
     def run(self, filename, file, timeout: int=DEFAULT_TIMEOUT):
-        '''
-        Yields results for all invalid links in a file.
+        """
```
Comment (Adrianzatreanu) [inline bears/general/InvalidLinkBear.py]: why exactly are we using this kind of docstring here?

File: bears/c_languages/ClangComplexityBear.py [hunk 12]
```diff
@@ -85,10 +85,20 @@ def run(self, filename, file, max_complexity: int=8):
                 affected_code = (SourceRange.from_clang_range(cursor.extent),)
                 yield Result(
                     self,
-                    "The cyclomatic complexity of function {function} is "
-                    "{complexity} which exceeded maximal recommended value "
+                    "The function '{function}' should be simplified. Its "
+                    "cyclomatic complexity is {complexity} which exceeds "
+                    "maximal recommended value "
```
Comment (Adrianzatreanu) [inline bears/c_languages/ClangComplexityBear.py]: ``maxim recommended value`` which is?

Comment (sils1297) [inline bears/c_languages/ClangComplexityBear.py]: see line below

**Comment by Adrianzatreanu** [comment]
ack 09794e5

**Comment by Adrianzatreanu** [comment]
ack 3d5cf61

File: bears/c_languages/ClangComplexityBear.py (hunk 12 above)
Comment (Adrianzatreanu) [inline bears/c_languages/ClangComplexityBear.py]: oh! you're right, i didnt see that.

**Comment by Adrianzatreanu** [comment]
ack 589b2ae

**Comment by Adrianzatreanu** [comment]
ack 77be9b8

**Comment by sils1297** [comment]
reack
5e6996f GitCommitBear: Use imperative tense
21c1413 SpaceConsistencyBear: Add additional info
103d755 ClangComplexityBear: Add result explanation
42815cb SpaceConsistencyBear: Extend documentation
acf3b0d InvalidLinkBear: Improve documentation
17ef349 GitCommitBearTest: Adapt to coala changes

**Comment by sils1297** [comment]
coverage in the circle build log is 100% so this should be mergeable (I guess codecov state still not appearing until we add a config)

**Comment by sils1297** [comment]
reack 64762a9 SpaceConsistencyBear: Add additional info
11760b8 ClangComplexityBear: Add result explanation
e4bf512 SpaceConsistencyBear: Extend documentation
17c1d9c InvalidLinkBear: Improve documentation
86eba62 GitCommitBearTest: Adapt to coala changes

**Comment by sils1297** [comment]
reack 4dbc637 GitCommitBear: Use imperative tense
40d8f59 SpaceConsistencyBear: Add additional info
8ba2c64 ClangComplexityBear: Add result explanation
77e8462 SpaceConsistencyBear: Extend documentation
6480db0 InvalidLinkBear: Improve documentation

**Comment by sils1297** [comment]
ack (self because hotfix) 4da7fd0

**Comment by sils1297** [comment]
@rultor merge
